org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/rename/LSPRenameParticipant.java

```diff
@@ -0,0 +1,124 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Dawid Pakuła and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Dawid Pakuła - initial implementation
+ *******************************************************************************/
+package org.eclipse.lsp4e.operations.rename;
+
+import java.nio.file.FileSystems;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.lsp4e.LSPEclipseUtils;
+import org.eclipse.lsp4e.LanguageServerWrapper;
+import org.eclipse.lsp4e.LanguageServers;
+import org.eclipse.lsp4e.internal.Pair;
+import org.eclipse.lsp4j.FileOperationOptions;
+import org.eclipse.lsp4j.FileRename;
+import org.eclipse.lsp4j.RenameFilesParams;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.eclipse.ltk.core.refactoring.Change;
+import org.eclipse.ltk.core.refactoring.CompositeChange;
+import org.eclipse.ltk.core.refactoring.RefactoringStatus;
+import org.eclipse.ltk.core.refactoring.participants.CheckConditionsContext;
+import org.eclipse.ltk.core.refactoring.participants.RenameParticipant;
+
+public class LSPRenameParticipant extends RenameParticipant {
+
+	private IFile file;
+	private List<Pair<LanguageServerWrapper, LanguageServer>> servers;
+
+	@SuppressWarnings("null")
+	static List<Pair<LanguageServerWrapper, LanguageServer>> collectServers(IFile file)
+	{
+		return LanguageServers.forProject(file.getProject()).withFilter(f -> {
+			if (f.getWorkspace() == null || f.getWorkspace().getFileOperations() == null) {
+				return false;
+			}
+			FileOperationOptions willRename = f.getWorkspace().getFileOperations().getWillRename();
+			if (willRename == null) {
+				return false;
+			}
+			if (willRename.getFilters() == null || willRename.getFilters().isEmpty()) {
+				return true;
+			}
+			return willRename.getFilters().stream().anyMatch(filter -> {
+				return FileSystems.getDefault().getPathMatcher("glob:" + filter.getPattern().getGlob()) //$NON-NLS-1$
```

Comment:
`FileSystems.getDefault().getPathMatcher("glob:" `have  a limitation, when you define `**/*.ts` and your base dir is 

`C:/Users/`

 * `C:/Users/foo/bar.ts`  will match `**/*.ts`
 * `C:/Users/bar.ts`  will NOT match `**/*.ts` although vscode support it and it sa very commons usecases.

More its is better to cache the globfile pattern.

I suggest that you see my test code at https://github.com/redhat-developer/lsp4ij/blob/main/src/test/java/com/redhat/devtools/lsp4ij/features/filewatchers/FileSystemWatcherManagerTest.java which are the same test than vscode.

In otherwords I suggest that you copy/paste my https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/features/filewatchers/PathPatternMatcher.java and use it.

PathPatternMatcher is used to manage glob pattern for didChangeFilesWatcher and willRenameFiles, etc

Here my current PR to support willRenameFiles https://github.com/redhat-developer/lsp4ij/pull/198

Hope it will help.